libp2p/network/swarm.py

```diff
@@ -238,14 +229,19 @@ def set_stream_handler(self, stream_handler: StreamHandlerFn) -> None:
                 logger.debug("upgraded mux for peer %s", peer_id)
 
                 await self.add_conn(muxed_conn)
-
                 logger.debug("successfully opened connection to peer %s", peer_id)
 
+                # NOTE: This is a intentional barrier to prevent from the handler exiting and
+                #   closing the connection.
+                await self.manager.wait_finished()
+
             try:
                 # Success
                 listener = self.transport.create_listener(conn_handler)
                 self.listeners[str(maddr)] = listener
-                await listener.listen(maddr)
+                # TODO: `listener.listen` is not bounded with nursery. If we want to be
+                #   I/O agnostic, we should change the API.
+                await listener.listen(maddr, self.manager._task_nursery)  # type: ignore
```

Comment:
We should not be accessing this private API.  Instead, this class should create a new nursery in the `run` method and that should be used here.